### PR TITLE
 JAMES-2766 Upgrading to ElasticSearch 6.7

### PR DESCRIPTION
**Pull request by chibenwa**
We propose to strategies:

 - A simple one that compromise search accuracy
 - Another one that minimize search inconsistencies but requires more operations.

### CODE REVIEW

File: upgrade-instructions.md [hunk 1]
```diff
@@ -14,6 +14,41 @@ Note: this section is in progress. It will be updated during all the development
...
 
+Change list:
+
+ - [Upgrade to ElasticSearch 6.7](#upgrade-to-elasticsearch-6.7)
+
+#### Upgrade to ElasticSearch 6.7
+
+In version 3.3.0 indexing for the Cassandra product was handled using ElasticSearch 2.2 released on the 31 march 2016. Some major upgrades had been included in recent ElasticSearch version.
+
+Note that ElasticSearch APIs had been undergoing some major changes, making a smooth migration hard to provide. We proposed 2 migration strategy. A
```
Comment (rouazana) [inline upgrade-instructions.md]: strategies

File: upgrade-instructions.md [hunk 2]
```diff
@@ -14,6 +14,41 @@ Note: this section is in progress. It will be updated during all the development
...
+##### Simple strategy
+
+Procedure:
+ - From a running James 3.3.0 cluster connected to a running ElacticSearch 2.2 cluster
+ - Start an empty ElasticSearch 6.7 cluster
+ - Shutdown James 3.3.0 cluster and start a James 3.4.0 cluster connected to ElasticSearch 6.7
+ - Search result will then be empty and thus innacurate
+ - Thus trigger a [Full ReIndexing](https://james.apache.org/server/manage-webadmin.html#ReIndexing_all_mails) to restore search consistency.
+
+Obviously this approach trades search consistency against ease of use.
```
Comment (rouazana) [inline upgrade-instructions.md]: s/ease of use/ease of migration/ ?

File: upgrade-instructions.md [hunk 3]
```diff
@@ -14,6 +14,41 @@ Note: this section is in progress. It will be updated during all the development
...
+
+If search consistency is important for you, consider the next approach
```
Comment (rouazana) [inline upgrade-instructions.md]: search consistency *during the migration*?

File: upgrade-instructions.md [hunk 4]
```diff
@@ -14,6 +14,41 @@ Note: this section is in progress. It will be updated during all the development
...
+
+##### Strategy for minimizing search inconsistency
```
Comment (rouazana) [inline upgrade-instructions.md]: idem?

File: upgrade-instructions.md [hunk 5]
```diff
@@ -14,6 +14,41 @@ Note: this section is in progress. It will be updated during all the development
...
+
+Procedure:
+ - From a running James 3.3.0 cluster connected to a running ElacticSearch 2.2 cluster
+ - Start an empty ElasticSearch 6.7 cluster
+ - Start a James 3.4.0 cluster connected to ElasticSearch 6.7 cluster as well as the Cassandra source of trust database. Traffic should be directed to the James 3.3.0 cluster.
+ - Trigger an offline [Full ReIndexing](https://james.apache.org/server/manage-webadmin.html#ReIndexing_all_mails) on the James 3.4.0 cluster
+ - Once done, direct the traffic to the James 3.4.0 cluster, and dispose the James 3.3.0 cluster as well as the ElasticSearch 2.2 cluster
+ - Search result will omit changes that took place during the switching process (starting from the reIndexing start)
+ - Thus trigger a [Full ReIndexing](https://james.apache.org/server/manage-webadmin.html#ReIndexing_all_mails) to restore search consistency.
```
Comment (rouazana) [inline upgrade-instructions.md]: maybe warn the full reindexing can be slow, dependending on the data size?

File: upgrade-instructions.md [hunk 6]
```diff
@@ -14,6 +14,41 @@ Note: this section is in progress. It will be updated during all the development
 
 Changes to apply between 3.3.x and 3.4.x will be reported here.
 
+Change list:
+
+ - [Upgrade to ElasticSearch 6.7](#upgrade-to-elasticsearch-6.7)
+
+#### Upgrade to ElasticSearch 6.7
```
Comment (rouazana) [inline upgrade-instructions.md]: is there a way to inform that it concerns only the (experimental) james distributed product?

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**